ingester/src/wal/disk_full_protection.rs

```diff
@@ -0,0 +1,505 @@
+pub(crate) struct WalPersister<T, P> {
+    wal: Arc<wal::Wal>,
+    wal_reference_handle: WalReferenceHandle,
+    buffer: T,
+    persist: P,
+}
+
+impl<T, P> WalPersister<T, P>
+where
+    T: PartitionIter + Sync,
+    P: PersistQueue + Clone,
+{
+    /// Create a new [`wal::Wal`] based [`PersistingCleaner`] that uses the given
+    /// [`WalReferenceHandle`] to control tidying up of fully persisted WAL
+    /// segments.
+    pub fn new(
+        wal: Arc<wal::Wal>,
+        wal_reference_handle: WalReferenceHandle,
+        buffer: T,
+        persist: P,
+    ) -> Self {
+        Self {
+            wal,
+            wal_reference_handle,
+            buffer,
+            persist,
+        }
+    }
+}
+
+#[async_trait]
+impl<T, P> PersistingCleaner for WalPersister<T, P>
+where
+    T: PartitionIter + Sync,
+    P: PersistQueue + Clone,
+{
+    /// Rotate the WAL, notifying the reference tracker and persisting the buffer-tree.
+    ///
+    /// This function waits for the set of inactive WAL files to be empty before returning.
+    async fn persist_and_tidy(&self) {
+        let (closed_segment, sequence_number_set) =
+            self.wal.rotate().expect("failed to rotate wal");
+        debug!(
+            closed_id = %closed_segment.id(),
+            segment_bytes = closed_segment.size(),
+            n_ops = sequence_number_set.len(),
+            "rotated wal to allow disk clean-up",
+        );
+
+        let empty_waker = self.wal_reference_handle.empty_inactive_notifier();
+
+        _ = self
+            .wal_reference_handle
+            .enqueue_rotated_file(closed_segment.id(), sequence_number_set)
+            .await
+            .await;
+
+        persist_partitions(self.buffer.partition_iter(), &self.persist).await;
+        debug!(closed_id = %closed_segment.id(), "partitions persisted to allow disk clean-up");
+
+        empty_waker.await;
+    }
+}
+
+/// Protect the ingester from the disk watched by `disk_snapshot_rx` filling up,
+/// marking the provided `ingest_state` with an error while persisting
+/// outstanding partitions until enough space is free to clear the [`IngestStateError`].
+///
+/// This task runs until it detects the [`Receiver`] is disconnected.
+pub(crate) async fn guard_disk_capacity<P>(
+    mut disk_snapshot_rx: Receiver<DiskSpaceSnapshot>,
+    ingest_state: Arc<IngestState>,
+    persisting_cleaner: P,
+) where
+    P: PersistingCleaner + Send + Sync + 'static,
+{
+    let mut cleanup_handle: Option<JoinHandle<()>> = Default::default();
+    let persisting_cleaner = Arc::new(persisting_cleaner);
+
+    // Listen for new disk space snapshots in a loop, assessing the usage ratio
+    // and taking protective action if needed.
+    loop {
+        // If the sender has disconnected, this task cannot do anything
+        // meaningful and is likely a signal that shutdown has been invoked
+        // - abort the task..
+        if disk_snapshot_rx.changed().await.is_err() {
+            return;
```

Comment:
```suggestion
            info!("stopping disk full protection task");
            return;
```

Lets make sure we know when it stops!